File: backend/models/panchang.py
"""
Pydantic response models for Panchang API endpoints.
Maps 1:1 to the frontend PanchangData TypeScript interface.
"""

from pydantic import BaseModel
from typing import Optional


class PillarData(BaseModel):
    """A single Panchang pillar (Tithi, Nakshatra, Yoga, Karana, Vaara)."""
    name: str
    sanskrit: str
    meaning: str
    value: str
    sanskritValue: str
    deity: Optional[str] = None
    endTime: Optional[str] = None
    progressPercent: float = 0.0
    note: str = ""


class SolarLunarData(BaseModel):
    """Solar and lunar cycle data."""
    sunrise: str
    sunset: str
    moonrise: str
    moonset: str
    solarNoon: str
    sunSign: str
    moonSign: str
    moonPhaseName: str
    moonPhaseSanskrit: str
    illuminationPercent: float
    lunarAgeDays: float
    paksha: str
    lunarMonth: str
    samvatYear: str
    samvatName: str
    ritu: str
    ayana: str


class DayCharacterData(BaseModel):
    """Interpretation of the day's character."""
    title: str = "The Character of Today"
    sanskritTitle: str = "अद्यतन स्वभाव एवं प्रवाह"
    description: str = ""
    auspiciousNote: str = ""
    color: str = ""
    element: str = ""


class PanchangResponse(BaseModel):
    """Complete Panchang response — matches frontend PanchangData interface."""
    date: str
    sanskritDate: str
    location: str
    vaara: PillarData
    tithi: PillarData
    nakshatra: PillarData
    yoga: PillarData
    karana: PillarData
    cycles: SolarLunarData
    dayCharacter: DayCharacterData

File: src/types/panchang.ts
export interface PanchangPillar {
  name: string;
  sanskrit: string;
  meaning: string;
  value: string;
  sanskritValue: string;
  deity?: string;
  endTime?: string;
  progressPercent: number; // 0 to 100
  note: string;
}

export interface SolarLunarCycle {
  sunrise: string;
  sunset: string;
  moonrise: string;
  moonset: string;
  solarNoon: string;
  sunSign: string;
  moonSign: string;
  moonPhaseName: string;
  moonPhaseSanskrit: string;
  illuminationPercent: number;
  lunarAgeDays: number;
  paksha: 'Shukla Paksha (शुक्ल पक्ष)' | 'Krishna Paksha (कृष्ण पक्ष)';
  lunarMonth: string;
  samvatYear: string;
  samvatName: string;
  ritu: string; // Season (e.g., Grishma / वर्षा)
  ayana: string; // Uttarayana / Dakshinayana
}

export interface PanchangData {
  date: string;
  sanskritDate: string;
  location: string;
  vaara: PanchangPillar; // Day
  tithi: PanchangPillar; // Tithi
  nakshatra: PanchangPillar; // Nakshatra
  yoga: PanchangPillar; // Yoga
  karana: PanchangPillar; // Karana
  cycles: SolarLunarCycle;
  dayCharacter: {
    title: string;
    sanskritTitle: string;
    description: string;
    auspiciousNote: string;
    color: string;
    element: string;
  };
}
